### PR TITLE
Add support for Resource Owner Password Credentials Grant

### PR DESCRIPTION
**Pull request by snowzach**
This PR adds support for Resource Owner Password Credentials Grant.
https://www.oauth.com/oauth2-servers/access-tokens/password-grant/

You enable it in the config by setting:
```
oauth2:
   passwordConnector: local
```
or replace local with whatever connector you wish to use for password grants.

You then request a token via a 

```
POST /dex/token
grant_type=password&client_id=example-app&client_secret=ZXhhbXBsZS1hcHAtc2VjcmV0&username=admin@example.com&password=password&scope=openid profile email&nonce=2OSwiYXRfaGFzaCI6Il9rMVBZT0ppcGFQc0pHSmlZV
```

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -132,6 +135,9 @@ type Server struct {
 	// If enabled, don't prompt user for approval after logging in through connector.
 	skipApproval bool
 
+	// Used for password grant
+	passwordConnector string
```
Comment (ericchiang) [inline server/server.go]: nit, this should be a `connector.PasswordConnector`

File: server/server.go [hunk 2]
```diff
@@ -67,6 +67,9 @@ type Config struct {
 	// Logging in implies approval.
 	SkipApprovalScreen bool
 
+	// If set, the server will use this connector to handle password grants
+	PasswordConnector string
```
Comment (ericchiang) [inline server/server.go]: Same here, the connector should be initialized in `cmd` so this would be of type `connector.PasswordConnector`

**Comment by pivotal-jwinters** [comment]
@snowzach are you still looking at this? Really excited for these changes

**Comment by snowzach** [comment]
@pivotal-jwinters I wanted to finish this quick but the request changes are a bigger refactor for the code than I have time at the moment. I will get to it eventually but $dayjob has taken over for the time being. Sorry. :(

**Comment by pivotal-jwinters** [comment]
@snowzach no worries. If there's anything I can do to help let me know.

**Comment by yadzhang** [comment]
@snowzach @ericchiang  Any progress on this?  I am really need on password grant.

**Comment by xtremerui** [comment]
Hi, @bonifaido. I went through recent PRs and found you are assigned most of them so I figured you might be the right one to talk with. Sorry in advance.

@pivotal-jwinters and I works for [ConcourseCI](https://github.com/concourse) who uses Dex as its auth backend. We are currently using our [own fork](https://github.com/concourse/dex) of Dex and `password grant` in this PR has been merged into our fork for almost one year without any issue.

We would love to help here. Could you review the current status of the PR and provide some guidance? @snowzach if you dont mind could we pick up your work? If necessary we could move on to a new PR of `password grant`, which will be based on @snowzach 's first commit.

Thx!

**Comment by snowzach** [comment]
@xtremerui by all means you may do anything you like this with merge. I kept meaning to come back to it but just never had time.

**Comment by bonifaido** [comment]
Hi @xtremerui ! I think this PR looks quite good, we just need to rebase it on current master and resolve the conflicts. Would you guys do that?

**Comment by xtremerui** [comment]
@bonifaido perfecto! We will update the PR and submit it again from our fork.

**Comment by xtremerui** [comment]
https://github.com/dexidp/dex/pull/1621 is submitted. This PR might be closed. Thx!

**Comment by snowzach** [comment]
Close in favor of #1621

**Pull request closed without merging**